### DISCUSSION TITLE
Is zk.nvim still needed?

### DISCUSSION THREAD

**Discussion by mickael-menu** [megalithic/zk.nvim#41] [General]
`zk.nvim` was created by @megalithic to interface [`zk`](https://github.com/mickael-menu/zk) with Neovim, back when `zk` didn't have an [LSP server](https://github.com/mickael-menu/zk/issues/22).

Today, things are quite different. The LSP server covers most use cases of `zk.nvim`, and the Neovim plugin doesn't have any active maintainer anymore. The duplicated features between `zk.nvim` and `zk` seem to be a source of confusion as well in the issues tracker.

So I'm considering archiving `zk.nvim`, but would like to get some community feedback first. Is there a feature provided by `zk.nvim` that can't be handled by the LSP server or another dedicated plugin (e.g. fzf or telescope for search)? For the list of stuff supported by the LSP server, [check out this documentation](https://github.com/mickael-menu/zk/blob/main/docs/editors-integration.md).

We could also consider creating a lighter plugin that is just setting up the LSP server and maybe providing some telescope hooks.

cc @pstuifzand @chris-sanders @eric-hansen @cormacrelf  @pinpox @entropitor @vod3 @frandsoh @thecontinium

**Comment by eric-hansen** [comment]
It's a good question and I wish I had an answer to it.  I think though that if `zk.nvim` isn't going to be maintained then by the original author it still warrants something from someone else.  Maybe as a fresh start since LSP is more prominent in Neovim?  

But seems ZK itself is more than capable of handling what an LSP does without a proxy plugin at this point, so it is probably better to move all of what this plugin does into a simple LSP wrapper and submit it to neovim/nvim-lspconfig.

I used to use Obsidian heavily then moved to Logseq but I enjoy ZK since it does one thing (backlink/PKM) and does it well.

I don't know Go so not sure how much I can contribute to the LSP but I can always try.

**Reply by thecontinium** [reply]
Wrapping everything into the LSP and submitting to neovim/nvim-lspconfig seems like a good approach along with some lightweight telescope mappings.

**Reply by eric-hansen** [reply]
I'd say either make it optional for telescope or have a custom `on_attach` method that sets the keybindings.  Not sure if it would be a good approach but it seems acceptable, since not everyone uses Telescope.

I really wouldn't mind doing the LSP aspect of things if people are okay with that.

**Reply by mickael-menu** [reply]
> so it is probably better to move all of what this plugin does into a simple LSP wrapper and submit it to neovim/nvim-lspconfig.

Yes that's what I had in mind. Though [I had a discussion](https://github.com/neovim/neovim/pull/14546#issuecomment-841550854) with one of the maintainer of nvim-lspconfig and from it I gather that it would be more suitable as an independent plugin.


> I'd say either make it optional for telescope or have a custom `on_attach` method that sets the keybindings. Not sure if it would be a good approach but it seems acceptable, since not everyone uses Telescope.

`zk.nvim` has an approach that could be convenient for this, you need to explicitly enable the telescope support with:

```lua
require('telescope').load_extension('zk')
```

> I really wouldn't mind doing the LSP aspect of things if people are okay with that.

Sure, contributions are most welcome! I would like to keep this project under my own username this time though, so that it's easier to ensure the continuity with maintainers.

This is my setup that could be used as a starting block, since it exposes the `ZkNew` and `ZkIndex` commands:


<details><summary><tt>~/.config/nvim/init.vim</tt></summary>

```lua
-- Declare some custom Neovim commands as shortcuts to the zk functions.
command! -nargs=0 ZkIndex :lua require'lspconfig'.zk.index()
command! -nargs=? ZkNew :lua require'lspconfig'.zk.new(<args>)

lua << EOF

local lspconfig = require'lspconfig'
local configs = require'lspconfig/configs'

-- Initialize the zk language server
configs.zk = {
  default_config = {
    cmd = {'zk', 'lsp', '--log', '/tmp/zk-lsp.log'};
    filetypes = {'markdown'};
    root_dir = lspconfig.util.root_pattern('.zk');
    settings = {};
  };
}

-- Index the notebook of the current note.
configs.zk.index = function()
  vim.lsp.buf.execute_command({
    command = "zk.index",
    arguments = {vim.api.nvim_buf_get_name(0)},
  })
end

-- Create a new note in the current notebook.
configs.zk.new = function(...)
  vim.lsp.buf_request(0, 'workspace/executeCommand',
    {
        command = "zk.new",
        arguments = {
            vim.api.nvim_buf_get_name(0),
            ...
        },
    },
    function(_, _, result)
      if not (result and result.path) then return end
      vim.cmd("edit " .. result.path)
    end
  )
end

lspconfig.zk.setup({
  on_attach = function(client, bufnr)
    -- Key mappings
    local function buf_set_keymap(...) vim.api.nvim_buf_set_keymap(bufnr, ...) end
    local opts = { noremap=true, silent=false }
    buf_set_keymap("i", "<S-tab>", "<cmd>lua vim.lsp.buf.completion()<CR>", opts)
    -- Follow a Markdown link with <CR>.
    buf_set_keymap("n", "<CR>", "<cmd>lua vim.lsp.buf.definition()<CR>", opts)
    -- Preview a note with K when the cursor is on a link.
    buf_set_keymap("n", "K", "<cmd>lua vim.lsp.buf.hover()<CR>", opts)
    -- Create a new note using the current visual selection for the note title. This will replace the selection with a link to the note.
    buf_set_keymap("v", "<CR>", ":'<,'>lua vim.lsp.buf.range_code_action()<CR>", opts)
    -- Reindex the notebook. Usually the language server does this automatically, so it's not often needed.
    buf_set_keymap("n", "<leader>zi", ":ZkIndex<CR>", opts)
    -- Create a new note after prompting for a title.
    buf_set_keymap("n", "<leader>zn", ":ZkNew {title = vim.fn.input('Title: ')}<CR>", opts)
    -- Create a new daily note in my `log/` notebook directory.
    buf_set_keymap("n", "<leader>zl", ":ZkNew {dir = 'log'}<CR>", opts)
    -- Find the backlinks for the note linked under the cursor.
    buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
  end
})

EOF
```

</details>

**Reply by eric-hansen** [reply]
> Sure, contributions are most welcome! I would like to keep this project under my own username this time though, so that it's easier to ensure the continuity with maintainers.

Makes sense.  Then do you want to make a repo for the LSP so I can fork it or how do you want to go about that?

**Reply by mickael-menu** [reply]
I created https://github.com/mickael-menu/zk-nvim and invited you as a collaborator. Feel free to push directly to the repository for now.

**Reply by megalithic** [reply]
Huge thank you to everyone that has contributed and taken over this project and kept pushing `zk` forward. @mickael-menu thank you for all of your amazing work.

**Reply by megalithic** [reply]
I'm happy to deprecate this repo and push users on over to https://github.com/mickael-menu/zk-nvim

**Reply by megalithic** [reply]
@mickael-menu - I've transferred ownership of this repo to you. Let me know if you'd rather me cancel it so that you might continue with that freshly created repo under your user.

**Reply by mickael-menu** [reply]
@megalithic And thank you for kickstarting the Neovim plugin, it was instrumental in the success of `zk`!

> I've transferred ownership of this repo to you. Let me know if you'd rather me cancel it so that you might continue with that freshly created repo under your user.

I think it's best to keep it there to avoid confusing people and also to let users that want to keep using `zk.nvim` a place to do so. Could you just archive it in the settings to make clear that it won't receive new updates? I'll update the `README` when the LSP config plugin is ready.

Thanks again!

**Reply by eric-hansen** [reply]
I'll start work on the LSP integration this week.  Shouldn't take long just need to get some stuff with work done first.
